fix: exporters/cmake/OpenSSLConfig.cmake.in to work for build config

levitte opened the pull request:
This template file is made to make both:

1. OpenSSLConfig.cmake (CMake config used when building a CMake package
   against an uninstalled OpenSSL build)
2. exporters/OpenSSLConfig.cmake (CMake config that's to be installed
   alongside OpenSSL, and is used when building a CMake package against
   an OpenSSL installation).

Variant 1 was unfortunately getting the internal '_ossl_prefix' variable
wrong, which is due to how the perl snippet builds the command(s) to figure
out its value.  That needed some correction.

-----

Additionally, changes were made to deal with the fact that the build directory
may be separate from the source directory, and that the CMake config used with
uninstalled builds needs to deal with two include directories as a result, not
just one.

levitte committed:
fix: exporters/cmake/OpenSSLConfig.cmake.in to work for build config

This template file is made to make both:

1. OpenSSLConfig.cmake (CMake config used when building a CMake package
   against an uninstalled OpenSSL build)
2. exporters/OpenSSLConfig.cmake (CMake config that's to be installed
   alongside OpenSSL, and is used when building a CMake package against
   an OpenSSL installation).

Variant 1 was unfortunately getting the internal '_ossl_prefix' variable
wrong, which is due to how the perl snippet builds the command(s) to figure
out its value.  That needed some correction.

vdukhovni commented on exporters/cmake/OpenSSLConfig.cmake.in:
```diff
@@ -91,11 +91,13 @@ get_filename_component(_ossl_prefix "${CMAKE_CURRENT_LIST_FILE}" PATH)
-      foreach (split '/', $d);
+  if ($d ne '.') {
+      $OUT .= 'get_filename_component(_ossl_prefix "${_ossl_prefix}" PATH)' . "\n"
+          foreach (split '/', $d);
```
But I'm curious about the `for (...)` here, which uses `$_` presumably as the iteration variable, which does not appear to be used (both originally, and in the improved code). How does this work?

levitte replied:
It is a statement modifier, which is also why it's indented like that compared to the line before (the statement it modifies).

https://perldoc.perl.org/perlsyn#Statement-Modifiers

levitte replied:
Ah, never mind, that's not what you asked.  Yeah, `$_` isn't used, the `foreach` loop is simple there to count the number of times that same CMake statement should be added to `$OUT`, based on the number of directories in `$d`.

levitte replied:
I could actually have this instead:

``` perl
      $OUT .= 'get_filename_component(_ossl_prefix "${_ossl_prefix}" PATH)' . "\n"
          x scalar (split '/', $d);
```

I hope that every reader knows the `x` operator 😉

Sashan reviewed (approved):
looks good to me.

levitte commented:
Hmmmm, I may have forgotten something.  Gotta check

levitte committed:
fix: for exporters to work for build config, there may be two include dirs

For CMake / pkg-config configuration files to be used for an uninstalled
build, the include directory in the build directory isn't enough, if that
one is separate from the source directory.  The include directory in the
source directory must be accounted for too.

This includes some lighter refactoring of util/mkinstallvars.pl, with the
result that almost all variables in builddata.pm and installdata.pm have
become arrays, even though unnecessarily for most of them; it was simpler
that way.  The CMake / pkg-config templates are adapted accordingly.

levitte commented:
Please review again

t8m commented on exporters/cmake/OpenSSLConfig.cmake.in:
```diff
@@ -129,13 +129,15 @@ set(OPENSSL_VERSION_FIX "${OpenSSL_VERSION_PATCH}")
-set(OPENSSL_MODULES_DIR "${_ossl_prefix}/{- unixify($OpenSSL::safe::installdata::LIBDIR_REL_PREFIX, 1); -}/{- unixify($OpenSSL::safe::installdata::MODULESDIR_REL_LIBDIR, 1); -}")
-set(OPENSSL_RUNTIME_DIR "${_ossl_prefix}/{- unixify($OpenSSL::safe::installdata::BINDIR_REL_PREFIX, 1); -}")
+set(OPENSSL_LIBRARY_DIR "${_ossl_prefix}/{- unixify($OpenSSL::safe::installdata::LIBDIR_REL_PREFIX[0], 1); -}")
+set(OPENSSL_INCLUDE_DIR{- $OUT = '';
+                          $OUT .= ' "${_ossl_prefix}/' . $_ . '"'
+                              foreach (@OpenSSL::safe::installdata::INCLUDEDIR_REL_PREFIX); -})
```
I am probably confused here - does this assume two include directories for the target installation? Two include dirs are needed only during the build of the OpenSSL itself. Once it is installed there should be only single one for apps to use. But maybe I am just confused by this being in installdata??

levitte replied:
Yes you are.  There are two variants of OpenSSLConfig.cmake:

- one that ends up in $BLDSRC, i.e. the top of the build tree, which is used by people who want to link directly against that build without having to install.
  That's what the CMake `export()` function provides.
- one that ends up in $BLDSRC/exporters, which is the variant that's actually getting installed in the appropriate CMake install directory.
  That's what the CMAKE `install(EXPORT)` function provides.

The one in $BLDSRC needs two include dirs, one being $BLDSRC/include and one being $SRCDIR/include

levitte replied:
See the description of this PR 😉

levitte replied:
So maybe I should rename `installdata` to something else?  `exportdata`, perhaps?

levitte commented:
Uhm...  ping?

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Merged to the master and 3.3 branches. Thank you.
